lib/phoenix_live_view/colocated_js.ex

```diff
@@ -0,0 +1,337 @@
+  special care is necessary when you need to import other files inside the colocated JS:
+
+  ```javascript
+  import { someFunction } from "some-dependency";
+  import somethingElse from "@/vendor/vendored-file";
+  ```
+
+  While dependencies from `node_modules` should work out of the box, you cannot simply refer to your
+  `assets/vendor` folder using a relative path. Instead, your bundler needs to be configured to handle
+  an alias like `@` to resolve to your local `assets` folder. This is configured by default in the
+  esbuild configuration for new Phoenix applications using `esbuild`'s [alias option](https://esbuild.github.io/api/#alias).
+
+  ## Options
+
+  Colocated JavaScript can be configured through the attributes of the `<script>` tag.
+  The supported attributes are:
+
+    * `name` - The name under which the default export of the script is available when importing
+      the manifest. This is required, even if you don't plan to access the exported values or the
+      script does not actually export anything.
+
+    * `key` - A key under which to namespace the export. This is used by `Phoenix.LiveView.ColocatedHook` to
+      nest all hooks under the `hooks` key. For example, you could set this to `web_components` for each colocated
+      script that defines a web component and then access all of them as `colocated.web_components` when importing
+      the manifest.
+
+    * `extension` - a custom extension to use when writing the extracted file. The default is `js`.
+
+    * `manifest` - a custom manifest file to use instead of the default `index.js`. For example,
+      `web_components.ts`.
+
+  """
+
+  @behaviour Phoenix.Component.MacroComponent
+
+  @impl true
+  def transform({"script", attributes, [text_content]} = _ast, meta) do
+    opts = Map.new(attributes)
+    validate_name!(opts)
+    data = extract(opts, text_content, meta)
+
+    # we always drop colocated JS from the rendered output
+    {:ok, "", data}
+  end
+
+  def transform(_ast, _meta) do
+    raise ArgumentError, "ColocatedJS can only be used on script tags"
+  end
+
+  defp validate_name!(opts) do
+    case opts do
+      %{"name" => name} when is_binary(name) ->
+        :ok
+
+      %{"name" => name} ->
+        raise ArgumentError,
+              "the name attribute of a colocated JS / hook must be a compile-time string. Got: #{inspect(name)}"
+
+      _ ->
+        raise ArgumentError, "missing required name attribute for ColocatedJS"
+    end
+  end
+
+  @doc false
+  def extract(opts, text_content, meta) do
+    # _build/dev/phoenix-colocated/otp_app/MyApp.MyComponent/line_no.js
+    target_path =
+      target_dir()
+      |> Path.join(to_string(current_otp_app()))
+      |> Path.join(inspect(meta.env.module))
+
+    filename_opts =
+      %{name: opts["name"]}
+      |> maybe_put_opt(opts, "key", :key)
+      |> maybe_put_opt(opts, "manifest", :manifest)
+
+    filename = "#{meta.line}.#{opts["extension"] || "js"}"
```

Comment:
we could theoretically have more than one in a single line, no? Perhaps a hash of the contents for the filename?